### ISSUE TITLE
About FLOPs calculation in Table 2

### ISSUE THREAD

**Issue by HubHop** [whai362/PVT#1]
Hi Wenhai, thanks for this great work. 

I have few questions about the FLOPs calculation in this paper. Previously I tested the DeiT models with [ptflops](https://github.com/sovrasov/flops-counter.pytorch), I got 2.51G, 9.20G, 35.13G FLOPs for DeiT-Tiny, DeiT-Small, DeiT-Base, respectively. 

B.T.W I also included the matrix multiplications in the self-attention layer, namely `q @ k` and `attn @ v`. I assume there is something wrong with my calculation, may I know how do you calculate FLOPs for your experiments? 

Thanks.

**Comment by whai362** [comment]
Hi, thanks for your attention!

I calculate the flops of matrix operation by:

def mha_flops(h, w, dim, num_heads):
    dim_h = dim / num_heads
    n = h * w
    f1 = n * dim_h * n * num_heads
    f2 = n * n * dim_h * num_heads
    return f1 + f2

def get_vit_flops(net, input_shape, patch_size):
    flops, params = get_model_complexity_info(net, input_shape, as_strings=False)
    _, H, W = input_shape
    stage = mha_flops(H // patch_size, W // patch_size,
                      net.blocks[0].attn.dim,
                      net.blocks[0].attn.num_heads) * len(net.blocks)
    flops += stage
    return flops_to_string(flops), params_to_string(params)

If something is wrong, please let me know.

**Comment by whai362** [comment]
As I know, ptflops output MACs, instead of FLOPs.
We use the tools in mmdet https://github.com/open-mmlab/mmdetection/blob/master/tools/analysis_tools/get_flops.py

**Comment by HubHop** [comment]
Thanks for your quick reply! 

Roughly speaking, FLOPs should be equal to MACs * 2, but it seems like MACs in ptflops is equal to FLOPs in mmdet. In my case, I multiply MACs by 2 so there is a difference. I can get the same result without that multiplication.

**Comment by whai362** [comment]
> Thanks for your quick reply!
> 
> Roughly speaking, FLOPs should be equal to MACs * 2, but it seems like MACs in ptflops is equal to FLOPs in mmdet. In my case, I multiply MACs by 2 so there is a difference. I can get the same result without that multiplication.

```
net = model.vit_small_patch16_224()
input_shape = (3, 224, 224)
flops, params = get_vit_flops(net, input_shape, 16)
print(flops)
```
This code will get the result of 9.88 GFLOPs.

vit_small_patch16_224 is from timm.
``get_model_complexity_info'' is from mmdet.

**Comment by xwjabc** [comment]
I think `mha_flops()` indeed measures the amount of MACs (multiply + add) instead of FLOPs. The `R50: 4.1G` in Table 2 matches the MACs from ptflops or thop.
```
def mha_flops(h, w, dim, num_heads):
    dim_h = dim / num_heads
    n = h * w
    f1 = n * dim_h * n * num_heads
    f2 = n * n * dim_h * num_heads
    return f1 + f2
```

**Comment by khawar512** [comment]
Hi @whai362 and @HubHop. Please help me, I am facing confusion in calculating FLOP. Am i right?

```
import torch
from ptflops import get_model_complexity_info
from vit_pytorch.vits_face import ViTs_face

with torch.cuda.device(0):
 net = ViTs_face(loss_type="CosFace",
 GPU_ID=0,
 num_class=61,
 image_size=224,
 patch_size=8,
 ac_patch_size=12,
 pad=4,
 dim=512,
 depth=20,
 heads=8,
 mlp_dim=2048,
 dropout=0.1,
 emb_dropout=0.1)
 macs, params = get_model_complexity_info(net, (3, 224, 224), as_strings=True,
 print_per_layer_stat=True, verbose=True)
 print('{:<30} {:<8}'.format('Computational complexity: ', macs))
 print('{:<30} {:<8}'.format('Number of parameters: ', params))


def mha_flops(h, w, dim, num_heads):
 dim_h = dim / num_heads
 n = h * w
 f1 = n * dim_h * n * num_heads
 f2 = n * n * dim_h * num_heads
 return f1 + f2


result = mha_flops(224, 224, 512, 8)
print(result)
```
Computational complexity:       49.56 
GMacNumber of parameters:           63.67 M 
MHA_FLOPS   2578054119424.0

What is mha_flops?
